internal/gengapic/mixins.go

```diff
@@ -56,65 +119,102 @@ func (g *generator) getMixinFiles() []*descriptor.FileDescriptorProto {
+
+		g.imports[pbinfo.ImportSpec{Name: "iampb", Path: "google.golang.org/genproto/googleapis/iam/v1"}] = true
 	}
 
-	return false
-}
+	if g.hasLocationMixin() {
 
-func getLocationsDescriptor() *descriptor.FileDescriptorProto {
-	return mixinFiles["google.cloud.location.Locations"][0]
+		p("locationsClient locationpb.LocationsClient")
+		p("")
+
+		g.imports[pbinfo.ImportSpec{Name: "locationpb", Path: "google.golang.org/genproto/googleapis/cloud/location"}] = true
+	}
 }
 
-func getLocationsMethods() []*descriptor.MethodDescriptorProto {
-	return getLocationsDescriptor().GetService()[0].GetMethod()
+// mixinStubsInit prints the stub intialization for the mixin gRPC stubs that are
+// configured to be generated. This is used in the factory method of a generated client.
+func (g *generator) mixinStubsInit() {
+	p := g.printf
+
+	if g.hasLROMixin() {
+		p("    operationsClient: longrunningpb.NewOperationsClient(connPool),")
+	}
+	if g.hasIAMPolicyMixin() {
+		p("    iamPolicyClient: iampb.NewIAMPolicyClient(connPool),")
+	}
+	if g.hasLocationMixin() {
+		p("    locationsClient: locationpb.NewLocationsClient(connPool),")
+	}
 }
 
-func getIAMPolicyDescriptor() *descriptor.FileDescriptorProto {
-	return mixinFiles["google.iam.v1.IAMPolicy"][0]
+// hasLROMixin is a convenience method for determining if the Operations mixin
+// should be generated.
+func (g *generator) hasLROMixin() bool {
+	return len(g.mixins["google.longrunning.Operations"]) > 0 && len(g.serviceConfig.GetApis()) > 1
 }
 
-func getIAMPolicyMethods() []*descriptor.MethodDescriptorProto {
-	return getIAMPolicyDescriptor().GetService()[0].GetMethod()
+// hasIAMPolicyMixin is a convenience method for determining if the IAMPolicy
+// mixin should be generated.
+func (g *generator) hasIAMPolicyMixin() bool {
+	return len(g.mixins["google.iam.v1.IAMPolicy"]) > 0 && !g.hasIAMPolicyOverrides
 }
 
-func getOperationsDescriptor() *descriptor.FileDescriptorProto {
-	return mixinFiles["google.longrunning.Operations"][0]
+// hasLocationixin is a convenience method for determining if the Locations
+// mixin should be generated.
+func (g *generator) hasLocationMixin() bool {
+	return len(g.mixins["google.cloud.location.Locations"]) > 0
 }
 
-func getOperationsMethods() []*descriptor.MethodDescriptorProto {
-	return getOperationsDescriptor().GetService()[0].GetMethod()
+// checkIAMPolicyOverrides determines if any of the given services define an
+// IAMPolicy RPC and sets the hasIAMpolicyOverrides generator flag if so. If set
+// to true, the IAMPolicy mixin will not be generated on any service client. This
+// is for backwards compatibility with existing IAMPolicy redefinitions.
+func (g *generator) checkIAMPolicyOverrides(servs []*descriptor.ServiceDescriptorProto) {
+	iam, hasMixin := g.mixins["google.iam.v1.IAMPolicy"]
+	if !hasMixin {
+		return
+	}
+
+	for _, s := range servs {
+		for _, iamMethod := range iam {
+			if hasMethod(s, iamMethod.GetName()) {
+				g.hasIAMPolicyOverrides = true
```

Comment:
Nit: we could return after after setting the value.

Reply:
Well, it is referenced in multiple places, and I don't want to have to execute this method multiple times (mainly because its best to do it with the full list of Services-to-generate, which isn't available in the scope of other call sites) so I saved the state in the generator its called with. Seems simpler that way.

Reply:
Yes, I agree; I'm saying that we can short circuit the loop and end if we set `hasIAMPolicyOverrides` to true because it's a single, sticky value.

Reply:
Oh I totally misread your comment. yes, good catch! Thanks